sno/pk_generation.py

```diff
@@ -0,0 +1,176 @@
+import functools
+from .serialise_util import (
+    json_pack,
+    json_unpack,
+)
+from collections.abc import Iterable
+from .schema import ColumnSchema, Schema
+
+
+class PkGeneratingImportSource:
+    """
+    Wrapper of ImportSource that makes it appear to have a primary key,
+    even though the delegate ImportSource does not.
+    """
```

Comment:
can we add a more detailed explanation of how the hashes/keys are stored and updated over time?

Reply:
Seems to be memory-limited, right — O(N) with number of features?

And based on all the features ever loaded in the repo, not just the current/previous revision?

Reply:
these only apply to imports, right? not later edits, assuming they're not a reimport?

And what's the path to dealing with (or blocking) the merging multiple imports/edits?

Reply:
> can we add a more detailed explanation of how the hashes/keys are stored and updated over time?

Done

Reply:
> Seems to be memory-limited, right — O(N) with number of features?

Yup that's right - 40 bytes plus change for every feature ever imported / reimported into this repo. So, should be on the order of 40MB for every million features. Features edited not as part of an import don't count towards this total.

Reply:
> these only apply to imports, right? not later edits, assuming they're not a reimport?
Correct

> And what's the path to dealing with (or blocking) the merging multiple imports/edits?

This isn't handled specifically by this PR - and its not the intended flow - but it actually works as well as any other metadata conflict: let's say you import from the same source twice, but in two different branches, independently - each import is made without reference to the other (but perhaps both reference an earlier import). Then you can merge the two imports together, but you can get (up to) 3 types of conflicts

- the schema file can conflict if either import added new columns, since they will have new UUID.
- the generated-pks metadata file can conflict, if the generated mapping from feature to primary-key wasn't exactly the same each time
- every time two different features were imported with the same primary key, this will conflict

The good news is: you can resolve any of these by existing mechanisms. `sno resolve dataset:meta:generated-pks.json: --with=theirs` will accept their version of this metadata file.
The bad news is: resolving metadata conflicts is generally incomplete, and not just for generated primary keys. Although you can pick a winner, you can't do anything more complicated (ie merge two files into one), and picking a certain schema as the winner has implications for every feature, but there is no code for showing those implications yet. This is because conflict resolution hasn't been revisited since schema changes were added.